livy/batch.py

```diff
@@ -0,0 +1,150 @@
+
+    :param url: The URL of the Livy server.
+    :param file: File containing the application to execute.
+    :param class_name: Application Java/Spark main class.
+    :param proxy_user: User to impersonate when starting the session.
+    :param jars: URLs of jars to be used in this session.
+    :param py_files: URLs of Python files to be used in this session.
+    :param files: URLs of files to be used in this session.
+    :param driver_memory: Amount of memory to use for the driver process (e.g.
+        '512m').
+    :param driver_cores: Number of cores to use for the driver process.
+    :param executor_memory: Amount of memory to use per executor process (e.g.
+        '512m').
+    :param executor_cores: Number of cores to use for each executor.
+    :param num_executors: Number of executors to launch for this session.
+    :param archives: URLs of archives to be used in this session.
+    :param queue: The name of the YARN queue to which submitted.
+    :param name: The name of this session.
+    :param spark_conf: Spark configuration properties.
+    :param echo: Whether to echo output printed in the remote session. Defaults
+        to ``True``.
+    :param check: Whether to raise an exception when a statement in the remote
+        session fails. Defaults to ``True``.
+    """
+
+    def __init__(
+            self,
+            url: str,
+            auth: Auth = None,
+            file: str = None,
+            class_name: str = None,
+            args: List[str] = None,
+            proxy_user: str = None,
+            jars: List[str] = None,
+            py_files: List[str] = None,
+            files: List[str] = None,
+            driver_memory: str = None,
+            driver_cores: int = None,
+            executor_memory: str = None,
+            executor_cores: int = None,
+            num_executors: int = None,
+            archives: List[str] = None,
+            queue: str = None,
+            name: str = None,
+            spark_conf: Dict[str, Any] = None,
+            echo: bool = True,
+            check: bool = True,
+    ) -> None:
+        self.client = LivyClient(url, auth)
+        self.file = file
+        self.class_name = class_name
+        self.args = args
+        self.proxy_user = proxy_user
+        self.jars = jars
+        self.py_files = py_files
+        self.files = files
+        self.driver_memory = driver_memory
+        self.driver_cores = driver_cores
+        self.executor_memory = executor_memory
+        self.executor_cores = executor_cores
+        self.num_executors = num_executors
+        self.archives = archives
+        self.queue = queue
+        self.name = name
+        self.spark_conf = spark_conf
+        self.echo = echo
+        self.check = check
+        self.batch_id: Optional[int] = None
+
+    def __enter__(self) -> "LivyBatch":
+        self.start()
+        return self
+
+    def __exit__(self, exc_type, exc_value, traceback) -> None:
+        self.kill()
```

Comment:
How do you expect the `LivyBatch` object to be used?

The usage of `LivySession` as a context manager enables it to be automatically closed when the code block exits, but I'm not sure this makes sense for a non-interactive batch session.

Would it make more sense to drop the context manager behaviour and have something that works like:

```python
batch = LivyBatch(...)
batch.start()  # Actually creates the batch in Livy
batch.wait()  # Waits for a state indicating the batch is finished (either successfully or through an error)
batch.logs  # A `@property` that fetches the latest log lines
```

Keeping `batch.kill()` seems like a valuable method available to the user if their batch gets stuck.

I'd suggest that  `start` does not wait for the batch to actually begin - this was necessary for the interactive session in `LivySession` but does not seem so valuable here.

Reply:
Makes sense, thanks!
Also, I kept the `state` property to leave the ability for the program to perform asynchronously.

Reply:
> Also, I kept the `state` property to leave the ability for the program to perform asynchronously.

Sounds like a good idea!